sdk/couchbase-core/src/orphan_reporter.rs

```diff
@@ -0,0 +1,220 @@
+use std::time::Duration;
+use tokio::sync::mpsc;
+use tokio::time::{interval_at, Instant, MissedTickBehavior};
+
+#[derive(Debug, Clone)]
+pub struct OrphanContext {
+    pub client_id: String,
+    pub local_addr: SocketAddr,
+    pub peer_addr: SocketAddr,
+}
+
+#[derive(Debug, Eq)]
+struct OrphanLogItem {
+    pub connection_id: String,
+    pub operation_id: String,
+    pub remote_socket: String,
+    pub local_socket: String,
+    pub server_duration: Duration,
+    pub total_server_duration: Duration,
+    pub operation_name: String,
+}
+
+impl Display for OrphanLogItem {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let obj = json!({
+            "last_server_duration_us": self.server_duration.as_micros() as u64,
+            "total_server_duration_us": self.total_server_duration.as_micros() as u64,
+            "operation_name": self.operation_name,
+            "last_local_id": self.connection_id,
+            "operation_id": self.operation_id,
+            "last_local_socket": self.local_socket,
+            "last_remote_socket": self.remote_socket,
+        });
+        write!(f, "{}", obj)
+    }
+}
+
+struct OrphanLogJsonEntry {
+    count: u64,
+    top_items: Vec<OrphanLogItem>,
+}
+
+impl Display for OrphanLogJsonEntry {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let bodies: Vec<String> = self.top_items.iter().map(|item| item.to_string()).collect();
+        write!(
+            f,
+            r#"{{"total_count":{},"top_requests":[{}]}}"#,
+            self.count,
+            bodies.join(",")
+        )
+    }
+}
+
+struct OrphanLogService(HashMap<String, OrphanLogJsonEntry>);
+
+impl Display for OrphanLogService {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut parts = Vec::new();
+        for (svc, entry) in &self.0 {
+            parts.push(format!(r#""{}":{}"#, svc, entry));
+        }
+        write!(f, "{{{}}}", parts.join(","))
+    }
+}
+
+// Once we have total_duration added we will order on that
+impl PartialEq for OrphanLogItem {
+    fn eq(&self, other: &Self) -> bool {
+        self.total_server_duration == other.total_server_duration
+    }
+}
+
+impl Ord for OrphanLogItem {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.total_server_duration.cmp(&other.total_server_duration)
+    }
+}
+
+impl PartialOrd for OrphanLogItem {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+// Orphan reporter - Currently only handles/receives KV orphans
+pub struct OrphanReporter {
+    enabled: bool,
+    sender: Option<mpsc::Sender<OrphanLogItem>>,
+}
+
+impl OrphanReporter {
+    pub fn new(config: OrphanReporterConfig) -> Self {
+        if !config.enabled {
+            return Self {
+                enabled: config.enabled,
+                sender: None,
+            };
+        }
+        let (tx, mut rx) = mpsc::channel::<OrphanLogItem>(100);
```

Comment:
Why 100?

Reply:
Fair, changed it to use a Arc<Mutex<State>>> with our orphans + count.